File: Algorithm/countingsort.py
def counting_sort(arr):
    if not arr:
        return []
    
    min_val = min(arr)
    max_val = max(arr)
    range_val = max_val - min_val + 1
    
    count = [0] * range_val
    for x in arr:
        count[x - min_val] += 1
    
    for i in range(1, range_val):
        count[i] += count[i - 1]
    
    output = [0] * len(arr)
    for i in range(len(arr) - 1, -1, -1):
        idx = arr[i] - min_val
        count[idx] -= 1
        output[count[idx]] = arr[i]
    
    return output


data = [4, -1, 3, 0, 3, -1, 2, 4, 0]
print("Original array:", data)
sorted_data = counting_sort(data)
print("Sorted array:", sorted_data)

File: Algorithm/countingsort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

// 计数排序，稳定，支持负数
std::vector<int> countingSort(const std::vector<int>& a) {
    if (a.empty()) return {};

    int n = static_cast<int>(a.size());
    int minv = *std::min_element(a.begin(), a.end());
    int maxv = *std::max_element(a.begin(), a.end());
    int range = maxv - minv + 1;

    std::vector<int> count(range, 0);
    for (int x : a) count[x - minv]++;

    // 前缀和：count[i] 表示 <= i 的数量
    for (int i = 1; i < range; ++i) count[i] += count[i - 1];

    std::vector<int> out(n);
    // 从后向前遍历以保持稳定性
    for (int i = n - 1; i >= 0; --i) {
        int idx = a[i] - minv;
        count[idx]--;
        out[count[idx]] = a[i];
    }
    return out;
}

int main() {
    std::vector<int> data = {4, -1, 3, 0, 3, -1, 2, 4, 0};
    std::cout << "原数组: ";
    for (int x : data) std::cout << x << ' ';
    std::cout << '\n';

    std::vector<int> sorted = countingSort(data);

    std::cout << "排序后: ";
    for (int x : sorted) std::cout << x << ' ';
    std::cout << '\n';
    return 0;
}